rmw_dds_common/test/test_graph_cache.cpp

```diff
@@ -0,0 +1,1066 @@
+    check_names_and_namespace(names, namespaces, nodes_names_and_namespaces);
+  }
+
+  EXPECT_EQ(nodes_names_and_namespaces.size(), graph_cache.get_number_of_nodes());
+
+  {
+    rmw_names_and_types_t names_and_types = rmw_get_zero_initialized_names_and_types();
+    graph_cache.get_names_and_types(
+      demangle_topic,
+      demangle_type,
+      &allocator,
+      &names_and_types);
+    check_names_and_types(names_and_types, topics_names_and_types);
+  }
+}
+
+void check_results_by_node(
+  const GraphCache & graph_cache,
+  const std::string & node_namespace,
+  const std::string & node_name,
+  const NamesAndTypes & readers_names_and_types = {},
+  const NamesAndTypes & writers_names_and_types = {},
+  DemangleFunctionT demangle_topic = identity_demangle,
+  DemangleFunctionT demangle_type = identity_demangle)
+{
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+
+  {
+    std::cout << "checking readers" << std::endl;
+    rmw_names_and_types_t names_and_types = rmw_get_zero_initialized_names_and_types();
+    graph_cache.get_reader_names_and_types_by_node(
+      node_name,
+      node_namespace,
+      demangle_topic,
+      demangle_type,
+      &allocator,
+      &names_and_types);
+    check_names_and_types(names_and_types, readers_names_and_types);
+  }
+
+  {
+    std::cout << "checking writers" << std::endl;
+    rmw_names_and_types_t names_and_types = rmw_get_zero_initialized_names_and_types();
+    graph_cache.get_writer_names_and_types_by_node(
+      node_name,
+      node_namespace,
+      demangle_topic,
+      demangle_type,
+      &allocator,
+      &names_and_types);
+    check_names_and_types(names_and_types, writers_names_and_types);
+  }
+}
+
+void
+check_results_by_topic(
+  const GraphCache & graph_cache,
+  const std::string topic_name,
+  size_t readers_count = 0,
+  size_t writers_count = 0)
+{
+  {
+    // Start with a value different from the expected one, to
+    // ensure that the function is actually doing something.
+    size_t count = readers_count + 1;
+    EXPECT_EQ(
+      graph_cache.get_reader_count(topic_name, &count),
+      RMW_RET_OK);
+    EXPECT_EQ(readers_count, count);
+  }
+
+  {
+    size_t count = writers_count + 1;
+    EXPECT_EQ(
+      graph_cache.get_writer_count(topic_name, &count),
+      RMW_RET_OK);
+    EXPECT_EQ(writers_count, count);
+  }
+}
+
+TEST(test_graph_cache, zero_initialized)
+{
+  GraphCache graph_cache;
+
+  check_results(graph_cache);
+  check_results_by_node(graph_cache, "some_namespace", "node/name");
+  check_results_by_topic(graph_cache, "some/topic/name");
+}
+
+rmw_gid_t
+gid_from_string(const std::string & str)
+{
+  rmw_gid_t gid = {};
+  EXPECT_LT(str.size(), RMW_GID_STORAGE_SIZE);
+  std::strncpy(
+    reinterpret_cast<char *>(gid.data),
+    str.c_str(),
+    RMW_GID_STORAGE_SIZE);
+  return gid;
+}
+
+// gid, topic name, type name
+using EntitiesInfo = std::vector<std::tuple<std::string, std::string, std::string, bool>>;
```

Comment:
@ivanpauno same as elsewhere about using a `struct`. I'm not a big fan of `std::tuple` or `std::pair`, neither is particularly readable and you need to resort to comments to fill the gap. And you can also do aggregate initialization with `struct`s (as long as the constructor is not `explicit`).

Reply:
Yeah, that's true. I like using `std::tuple` and `std::pair` in test :rofl: .